lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs

```diff
@@ -163,152 +179,250 @@ parseGenesisData = \case
 -- For Integration
 --------------------------------------------------------------------------------
 
-data CardanoNodeConfig = CardanoNodeConfig
-    { nodeConfigFile   :: FilePath
-    , nodeDatabaseDir  :: FilePath
-    , nodeSocketFile   :: FilePath
-    , nodeTopologyFile :: FilePath
-    , nodeVrfKey       :: FilePath
-    , nodeKesKey       :: FilePath
-    , nodeOpCert       :: FilePath
-    }
-
--- | Spins up a @cardano-node@ in another process.
+-- | A quick helper to interact with the 'cardano-cli'. Assumes the cardano-cli
+-- is available in PATH.
+cli :: [String] -> IO String
+cli args =
+    readProcess "cardano-cli" args stdin
+  where
+    stdin = ""
+
+-- | Execute an action after starting a cluster of stake pools. The cluster also
+-- contains a single BFT node that is pre-configured with keys available in the
+-- test data.
 --
--- IMPORTANT: @cardano-node@ must be available on the current path.
-withCardanoNode
+-- This BFT node is essential in order to bootstrap the chain and allow
+-- registering pools. Passing `0` as a number of pool will simply start a single
+-- BFT node.
+withCluster
     :: Trace IO Text
-    -- ^ Some trace for logging
-    -> FilePath
-    -- ^ Test directory
+    -- ^ Trace for subprocess control logging
     -> Severity
+    -- ^ Minimal logging severity
+    -> Int
+    -- ^ How many pools should the cluster spawn.
     -> (FilePath -> Block -> (NetworkParameters, NodeVersionData) -> IO a)
-    -- ^ Callback function with a socket description and genesis params
-    -> IO a
-withCardanoNode tr tdir severity action =
-    orThrow $ withConfig tdir severity $ \cfg block0 (np, vData) -> do
-        nodePort <- getRandomPort
-        let args = mkArgs cfg nodePort
+    -- ^ Action to run with the cluster up
+    -> IO (Either ProcessHasExited a)
+withCluster tr severity n action = do
+    ports <- randomUnusedTCPPorts (n + 1)
+    print $ permutations ports
```

Comment:
probably we want to use our logging infrastructure someday in the future - if not doing this now maybe add TO_DO to not miss it later

Reply:
What do you mean?

Reply:
I mean, we have now `print` - and maybe use our logging here?

Reply:
Also, something more descriptive than
> [(32403,[])]

would be nice

Reply:
Ah sorry. I was really confused by your comment at first and didn't get why. This line is actually an oversight. I removed it :+1: